File: frontend/src/schemas/audit.ts
import { z } from "zod";
import { ledgerEventSchema, runLedgerCursorSchema } from "./ledger";
import { runStatusSchema } from "./invoice_read";

export const auditRunPageSchema = z.object({
  run_id: z.string().uuid(),
  invoice_id: z.string().uuid(),
  trace_id: z.string().regex(/^[0-9a-f]{32}$/),
  status: runStatusSchema,
  events: z.array(ledgerEventSchema).max(200),
  next_cursor: runLedgerCursorSchema.nullable(),
}).strict();

export type AuditRunPage = z.infer<typeof auditRunPageSchema>;

File: src/invoiceops_agent/api/schemas/audit.py
"""Paginated, role-protected run audit history."""

from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from invoiceops_agent.api.schemas.invoice_read import RunStatus
from invoiceops_agent.ledger.schemas import LedgerEvent, RunCursor, TraceId


class AuditRunPage(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True, allow_inf_nan=False)

    run_id: UUID
    invoice_id: UUID
    trace_id: TraceId
    status: RunStatus
    events: list[LedgerEvent] = Field(max_length=200)
    next_cursor: RunCursor | None
